chalk-parse/src/ast.rs

```diff
@@ -212,6 +212,10 @@ pub enum Ty {
     Slice {
         ty: Box<Ty>,
     },
+    Array {
+        ty: Box<Ty>,
+        len: Box<GenericArg>,
```

Comment:
Why is one here a `Ty` and the other ` GenericArg`?

Reply:
Yeah I'm not super pleased with it as is.
Basically the idea was that the first parameter has to be a type, hence it's a `Ty`, but for the array length you probably want to be able to use either a literal const value *or* a const parameter (i.e. either a `GenericArg::Id` or `GenericArg::ConstValue` in the AST). It seemed like the easiest way to abstract over those two possibilities was just to make the length a `GenericArg`.

Any ideas for a clean solution?

Reply:
Hmm let's see what @Areredify thinks

Reply:
So, I fell like the "idiomatic" solution in the spirit of the current implementation would be a) adding `ConstWithoutId` which is (for now) just `ConstValue` b) adding `Const` = `ConstWithoutId` + `Id` c) change `GenericArg::ConstValue` to `GenericArg::ConstWithoutId`. Then array would be `struct Array { ty: Box<Ty>, len: Box<Const> }`.
I can't really think of a better way to handle this without a handwritten parser

Reply:
Yeah that's along the lines of what I was thinking. I added a commit with those changes.